src/handlers/gestures/GestureMonitor.tsx

```diff
@@ -1,93 +1,399 @@
+  preparedGesture.animatedEventHandler = event;
+  preparedGesture.animatedHandlers = sharedHandlersCallbacks;
+}
+
+interface GestureMonitorProps {
+  gesture?: InteractionBuilder | GestureType;
+  animatedGesture?: InteractionBuilder | GestureType;
+}
+export const GestureMonitor: React.FunctionComponent<GestureMonitorProps> = (
+  props
+) => {
+  const useAnimated = props.animatedGesture != null;
+  const gestureConfig = props.animatedGesture ?? props.gesture;
+  const gesture = gestureConfig?.configure?.() ?? [];
+  const viewRef = useRef(null);
+  const firstRenderRef = useRef(true);
+
+  const preparedGesture = React.useRef<GestureConfigReference>({
+    config: gesture,
+    animatedEventHandler: null,
+    animatedHandlers: null,
+    firstExecution: true,
+    useAnimated: useAnimated,
+  }).current;
+
+  if (useAnimated !== preparedGesture.useAnimated) {
+    throw new Error(
+      'You cannot change whether you are using gesture or animatedGesture while the app is running'
+    );
   }
 
-  render() {
-    if (this.props.gesture?.animatedEventHandler) {
-      return (
-        <AnimatedWrap
-          ref={this.refHandler}
-          onGestureHandlerEvent={this.props.gesture?.animatedEventHandler}>
-          {this.props.children}
-        </AnimatedWrap>
-      );
+  if (preparedGesture.firstExecution) {
+    gestureConfig?.initialize?.();
+  }
+
+  if (useAnimated) {
+    // Whether animatedGesture or gesture is used shouldn't change
+    // during while an app is running
+    // eslint-disable-next-line react-hooks/rules-of-hooks
+    useAnimatedGesture(preparedGesture);
+  }
+
+  useEffect(() => {
+    firstRenderRef.current = true;
+    const viewTag = findNodeHandle(viewRef.current) as number;
+    attachHandlers(
+      preparedGesture,
+      gestureConfig,
+      gesture,
+      viewTag,
+      useAnimated
+    );
+
+    return () => {
+      dropHandlers(preparedGesture);
+    };
+  }, []);
+
+  useEffect(() => {
+    if (!firstRenderRef.current) {
+      const viewTag = findNodeHandle(viewRef.current) as number;
+
+      if (needsToReattach(preparedGesture, gesture)) {
+        dropHandlers(preparedGesture);
+        attachHandlers(
+          preparedGesture,
+          gestureConfig,
+          gesture,
+          viewTag,
+          useAnimated
+        );
+      } else {
+        updateHandlers(preparedGesture, gestureConfig, gesture);
+      }
     } else {
-      return React.Children.only(this.props.children);
+      firstRenderRef.current = false;
     }
+  }, [props]);
+
+  if (props.animatedGesture) {
+    return (
+      <AnimatedWrap
+        ref={viewRef}
+        onGestureHandlerEvent={preparedGesture.animatedEventHandler}>
+        {props.children}
+      </AnimatedWrap>
+    );
+  } else {
+    return <Wrap ref={viewRef}>{props.children}</Wrap>;
   }
-}
+};
 
-class Wrap extends React.Component<{ onGestureHandlerEvent: unknown }> {
+class Wrap extends React.Component<{ onGestureHandlerEvent?: unknown }> {
```

Comment:
Is it unknown though?

Reply:
In this case it is the result of the `useEvent` hook from `react-native-reanimated` and it's not documented (at least I didn't find anything useful), so kind of?

Reply:
https://github.com/software-mansion/react-native-reanimated/blob/3f1ac80d95fbf2736e3747a219a9e2afe133a7a7/src/reanimated2/hook/utils.ts#L14-L18

It's typed but we can leave it for now.